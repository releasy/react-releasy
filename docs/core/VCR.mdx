---
name: VCR
menu: Core
route: /core/vcr
---

# VCR

A class able to record and replay fetching results.

Usefull for testing purpouses, it will record all HTTP requests and replay on the later ones.

## Usage

```javascript
import { VCR } from 'react-releasy';

const link = new VCR({
  mode: VCR.MODE.AUTO,
  url: 'https://yourserveraddress.com/graphql',
});
```

## Why should I use VCR?

- Faster (no real HTTP requests are made anymore)
- Deterministic (the test will continue to pass, even if you are offline, or if the server goes down for maintenance)
- Accurate (the response will contain the same data you get from a real request)

## Modes

- **VCR.MODE.AUTO**: Record on first execution, replay the later ones.
- **VCR.MODE.RECORD**: Record all executions. (Overwrite data)
- **VCR.MODE.REPLAY**: Replay all executions.
