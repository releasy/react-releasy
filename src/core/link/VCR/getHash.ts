import { RequestNode, Variables } from 'relay-runtime';
import stableCopy from 'relay-runtime/lib/stableCopy.js';
import stringHash from 'string-hash';

const getHash = (request: RequestNode, variables: Variables): string => {
  const copy = stableCopy({
    variables,
    query: request.text,
  });

  const { name } = request;
  const hash = stringHash(JSON.stringify(copy)).toString();

  return `${name}_${hash}`;
};

export default getHash;
