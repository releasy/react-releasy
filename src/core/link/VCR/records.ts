import fse from 'fs-extra';
import { RequestNode, Variables, GraphQLResponse } from 'relay-runtime';
import stableCopy from 'relay-runtime/lib/stableCopy.js';
import stringHash from 'string-hash';

const RECORDS_FOLDER = './__vcr__';

const getPath = (request: RequestNode, variables: Variables): string => {
  const copy = stableCopy({
    variables,
    query: request.text,
  });

  const { name } = request;
  const hash = stringHash(JSON.stringify(copy)).toString();

  return `${RECORDS_FOLDER}/${name}_${hash}.json`;
};

export const clearRecords = (): Promise<void> => {
  return fse.emptyDir(RECORDS_FOLDER);
};

export const hasExistingRecord = (request: RequestNode, variables: Variables): Promise<boolean> => {
  const path = getPath(request, variables);

  return fse.pathExists(path);
};

export const readRecord = (
  request: RequestNode,
  variables: Variables,
): Promise<GraphQLResponse> => {
  const path = getPath(request, variables);

  return fse.readJson(path);
};

export const writeRecord = async (
  request: RequestNode,
  variables: Variables,
  data: GraphQLResponse,
): Promise<void> => {
  const path = getPath(request, variables);

  fse.ensureDir(RECORDS_FOLDER);

  return fse.writeJson(path, data);
};
