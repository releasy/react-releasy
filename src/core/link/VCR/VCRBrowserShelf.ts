import storage from 'local-storage-fallback';
import { GraphQLResponse } from 'relay-runtime';

import { VCRShelfInterface } from './VCR';

class VCRBrowserShelf implements VCRShelfInterface {
  get = (key: string): Promise<GraphQLResponse> => {
    return new Promise((resolve) => {
      resolve(JSON.parse(storage.getItem(key)));
    });
  }

  set = (key: string, data: GraphQLResponse): Promise<void> => {
    return new Promise((resolve) => {
      resolve(storage.setItem(key, JSON.stringify(data)));
    });
  }

  exists = (key: string): Promise<boolean> => {
    return new Promise((resolve, reject) => {
      resolve(!!storage.getItem(key));
    });
  }
}

export default VCRBrowserShelf;
