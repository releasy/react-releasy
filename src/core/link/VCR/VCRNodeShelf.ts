import fs from 'fs';
import path from 'path';
import { GraphQLResponse } from 'relay-runtime';

import { VCRShelfInterface } from './VCR';

export const DIR = path.join(__dirname, '__vcr__');

class VCRNodeShelf implements VCRShelfInterface {
  constructor() {
    if (!fs.existsSync(DIR)) {
      fs.mkdirSync(DIR);
    }
  }

  get = (key: string): Promise<GraphQLResponse> => {
    return new Promise((resolve, reject) => {
      const fullPath = path.join(DIR, key);

      fs.readFile(fullPath, 'utf8', (error, data) => {
        if (error) {
          return reject(error);
        }

        resolve(JSON.parse(data));
      });
    });
  }

  set = (key: string, data: GraphQLResponse): Promise<void> => {
    return new Promise((resolve, reject) => {
      const fullPath = path.join(DIR, key);

      fs.writeFile(fullPath, JSON.stringify(data), 'utf8', (error) => {
        if (error) {
          return reject(error);
        }

        resolve();
      });
    });
  }

  exists = (key: string): Promise<boolean> => {
    return new Promise((resolve, reject) => {
      const fullPath = path.join(DIR, key);

      resolve(fs.existsSync(fullPath));
    });
  }
}

export default VCRNodeShelf;
