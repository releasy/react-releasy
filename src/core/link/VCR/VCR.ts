import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';

import { LinkInterface } from '../../Config';
import Link, { LinkProps } from '../Link/Link';
import { hasExistingRecord, readRecord, writeRecord } from './records';

enum VCRMode {
  AUTO = 'Auto',
  RECORD = 'Record',
  REPLAY = 'Replay',
}

type VCRProps = LinkProps & {
  mode?: VCRMode,
};

class VCR implements LinkInterface {
  static MODE = VCRMode;

  mode: VCRMode;
  link: Link;

  constructor({ mode, ...linkProps }: VCRProps) {
    this.mode = mode || VCRMode.AUTO;
    this.link = new Link(linkProps);
  }

  recordMode = async (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<GraphQLResponse> => {
    const data = await this.link.fetch(request, variables, cacheConfig, uploadables);

    writeRecord(request, variables, data);

    return data;
  }

  replayMode = async (
    request: RequestNode,
    variables: Variables,
  ): Promise<GraphQLResponse> => {
    const data = await readRecord(request, variables);

    writeRecord(request, variables, data);

    return data;
  }

  autoMode = async (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<GraphQLResponse> => {
    const hasData = await hasExistingRecord(request, variables);
    if (hasData) {
      return this.replayMode(request, variables);
    }

    return this.recordMode(request, variables, cacheConfig, uploadables);
  }

  fetch = async (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<GraphQLResponse> => {
    if (this.mode === VCRMode.RECORD) {
      return this.recordMode(request, variables, cacheConfig, uploadables);
    }

    if (this.mode === VCRMode.REPLAY) {
      return this.replayMode(request, variables);
    }

    return this.autoMode(request, variables, cacheConfig, uploadables);
  }
}

export default VCR;
