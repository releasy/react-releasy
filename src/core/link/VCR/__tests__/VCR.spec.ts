import {
  createCacheConfig,
  createQueryRequest,
  createQueryResponse,
  createVariables,
} from '../../../../../tests/fixtures/relay';

import { clearRecords, readRecord } from '../records';
import VCR from '../VCR';

beforeEach(clearRecords);

it('should create a valid default VCR', () => {
  const vcr = new VCR({
    url: '/graphql',
  });

  expect(vcr).toMatchSnapshot();
});

it('should save data on record mode and retrieve data on replay mode', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const request = createQueryRequest();
  const variables = createVariables();
  const cacheConfig = createCacheConfig();

  const vcrWithRecordMode = new VCR({
    url: '/graphql',
    mode: VCR.MODE.RECORD,
  });

  expect(vcrWithRecordMode).toMatchSnapshot();

  const recordFetchData = await vcrWithRecordMode.fetch(request, variables, cacheConfig);
  const recordCacheData = await readRecord(request, variables);

  expect(recordFetchData).toEqual(recordCacheData);
  expect(recordCacheData).toMatchSnapshot();

  const vcrWithReplayMode = new VCR({
    url: '/graphql',
    mode: VCR.MODE.REPLAY,
  });

  expect(vcrWithReplayMode).toMatchSnapshot();

  const replayFetchData = await vcrWithReplayMode.fetch(request, variables, cacheConfig);

  expect(recordFetchData).toEqual(replayFetchData);
  expect(replayFetchData).toMatchSnapshot();
});

it('should save and retrieve data on auto mode', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const request = createQueryRequest();
  const variables = createVariables();
  const cacheConfig = createCacheConfig();

  const vcr = new VCR({
    url: '/graphql',
  });

  const recordFetchData = await vcr.fetch(request, variables, cacheConfig);
  const replayFetchData = await vcr.fetch(request, variables, cacheConfig);

  expect(recordFetchData).toEqual(replayFetchData);
  expect(replayFetchData).toMatchSnapshot();
});
