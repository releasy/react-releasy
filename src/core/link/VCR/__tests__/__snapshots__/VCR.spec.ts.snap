// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a valid default VCR 1`] = `
VCR {
  "autoMode": [Function],
  "fetch": [Function],
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "mode": "Auto",
  "recordMode": [Function],
  "replayMode": [Function],
}
`;

exports[`should save and retrieve data on auto mode 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "fakeViewerId",
      "name": "John Doe",
    },
  },
}
`;

exports[`should save data on record mode and retrieve data on replay mode 1`] = `
VCR {
  "autoMode": [Function],
  "fetch": [Function],
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "mode": "Record",
  "recordMode": [Function],
  "replayMode": [Function],
}
`;

exports[`should save data on record mode and retrieve data on replay mode 2`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "fakeViewerId",
      "name": "John Doe",
    },
  },
}
`;

exports[`should save data on record mode and retrieve data on replay mode 3`] = `
VCR {
  "autoMode": [Function],
  "fetch": [Function],
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "mode": "Replay",
  "recordMode": [Function],
  "replayMode": [Function],
}
`;

exports[`should save data on record mode and retrieve data on replay mode 4`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "fakeViewerId",
      "name": "John Doe",
    },
  },
}
`;
